src/misc/messages.cpp

```diff
@@ -62,47 +83,86 @@ class Message {
 		if (rendered_msg.empty())
 			rendered_msg = convert_ansi_markup(markup_msg.c_str());
 		assert(rendered_msg.length());
-		return rendered_msg.c_str();
+
+		if (!is_utf8_msg)
+			return rendered_msg.c_str();
+
+		const uint16_t cp = UTF8_GetCodePage();
+		if (rendered_cp_msg[cp].empty()) {
+			UTF8_RenderForDos(rendered_msg, rendered_cp_msg[cp], cp);
+			assert(rendered_cp_msg[cp].length());
+		}
+		return rendered_cp_msg[cp].c_str();
 	}
 
-	void Set(const char *markup)
+	void Set(const char *markup, const bool is_utf8)
 	{
 		assert(markup);
-		markup_msg = markup;
+		markup_msg  = markup;
+		is_utf8_msg = is_utf8;
+
 		rendered_msg.clear();
+		markup_cp_msg.clear();
+		rendered_cp_msg.clear();
 	}
 };
 
 static std::unordered_map<std::string, Message> messages;
 static std::deque<std::string> messages_order;
 
 // Add the message if it doesn't exist yet
-void MSG_Add(const char *name, const char *markup_msg)
+void MSG_Add(const char *name, const char *markup_msg, bool is_utf8)
 {
-	const auto &pair = messages.try_emplace(name, markup_msg);
+	const auto &pair = messages.try_emplace(name, markup_msg, is_utf8);
 	if (pair.second) // if the insertion was successful
 		messages_order.emplace_back(name);
 }
 
+void MSG_Add(const char *name, const char *markup_msg)
+{
+	MSG_Add(name, markup_msg, false);
+}
+
 // Replace existing or add if it doesn't exist
-void MSG_Replace(const char *name, const char *markup_msg)
+void MSG_Replace(const char *name, const char *markup_msg, bool is_utf8)
```

Comment:
Can you ditch the flag approach and introduce `MSG_Replace_Utf8`?

Reply:
I got rid of the extended function. It is actually not needed outside of messages.cpp, and I assume the plan is to migrate to gettext nevertheless.

Reply:
This is actually used internally only, it should be static.